src/types/YWeakLink.js

```diff
@@ -0,0 +1,206 @@
+    super()
+    this._id = id
+    this._linkedItem = item
+  }
+  
+  /**
+   * Returns a reference to an underlying value existing somewhere on in the document.
+   * 
+   * @return {T|undefined}
+   */
+  deref() {
+    if (this._linkedItem !== null && this._linkedItem.constructor === Item) {
+      let item = this._linkedItem
+      if (item.parentSub !== null) {
+        // for map types go to the most recent one
+        while (item.right !== null) {
+          item = item.right
+        }
+        this._linkedItem = item
+      }
+      if (!item.deleted) {
+        return item.content.getContent()[0]
+      }
+    }
+    return undefined;
+  }
+
+  /**
+   * Integrate this type into the Yjs instance.
+   *
+   * * Save this struct in the os
+   * * This type is sent to other client
+   * * Observer functions are fired
+   *
+   * @param {Doc} y The Yjs instance
+   * @param {Item|null} item
+   */
+  _integrate (y, item) {
+    super._integrate(y, item)
+    if (item !== null) {
+      transact(y, (transaction) => {
+        // link may refer to a single element in multi-element block
+        // in such case we need to cut of the linked element into a
+        // separate block
+        let sourceItem = this._linkedItem !== null ? this._linkedItem : getItemCleanStart(transaction, this._id)
+        if (sourceItem.constructor === Item && sourceItem.parentSub !== null) {
+          // for maps, advance to most recent item
+          while (sourceItem.right !== null) {
+            sourceItem = sourceItem.right
+          }
+        }
+        if (!sourceItem.deleted && sourceItem.length > 1) {
+          sourceItem = getItemCleanEnd(transaction, transaction.doc.store, createID(sourceItem.id.client, sourceItem.id.clock + 1))
+        }
+        this._linkedItem = sourceItem
+        if (!sourceItem.deleted) {
+          const src = /** @type {Item} */ (sourceItem)
+          if (src.linkedBy === null) {
+            src.linkedBy = new Set()
+          }
+          src.linkedBy.add(this)
+        }
+      })
+    }
+  }
+
+  /**
+   * @return {YWeakLink<T>}
+   */
+  _copy () {
+    return new YWeakLink(this._id, this._linkedItem)
+  }
+
+  /**
+   * @return {YWeakLink<T>}
+   */
+  clone () {
+    return new YWeakLink(this._id, this._linkedItem)
+  }
+
+  /**
+   * Creates YWeakLinkEvent and calls observers.
+   *
+   * @param {Transaction} transaction
+   * @param {Set<null|string>} parentSubs Keys changed on this type. `null` if list was modified.
+   */
+  _callObserver (transaction, parentSubs) {
+    super._callObserver(transaction, parentSubs)
+    callTypeObservers(this, transaction, new YWeakLinkEvent(this, transaction))
+  }
+
+  /**
+   * @param {UpdateEncoderV1 | UpdateEncoderV2} encoder
+   */
+  _write (encoder) {
+    encoder.writeTypeRef(YWeakLinkRefID)
+    const flags = 0 // flags that could be used in the future
+    encoding.writeUint8(encoder.restEncoder, flags)
+    encoder.writeLeftID(this._id)
```

Comment:
Encoding scheme is basically: 
1. `0` (1 byte) - flag that can be used in the future (eg. for links spanning over ranges rather than single elements etc.)
2. ID of linked element.

I'm not sure about `writeLeftID` here, as it will almost surely mess up with lib0 v2 delta-of-deltas encoding. Maybe just write ID into restEncoder?

Reply:
Yeah, I think that would be better